<!DOCTYPE html>

<html>
<head>
<meta content="text/html; charset=UTF-8" http-equiv="Content-Type">

<title>Table of Contents</title>

<link type="text/css" media="screen" href="./rdoc.css" rel="stylesheet">

<script type="text/javascript">
  var rdoc_rel_prefix = "./";
</script>

<script type="text/javascript" charset="utf-8" src="./js/jquery.js"></script>
<script type="text/javascript" charset="utf-8" src="./js/navigation.js"></script>
<script type="text/javascript" charset="utf-8" src="./js/search_index.js"></script>
<script type="text/javascript" charset="utf-8" src="./js/search.js"></script>
<script type="text/javascript" charset="utf-8" src="./js/searcher.js"></script>
<script type="text/javascript" charset="utf-8" src="./js/darkfish.js"></script>


<body class="indexpage">
<h1>Table of Contents</h1>

<h2>Pages</h2>
<ul>
  <li class="file">
    <a href="LICENSE_txt.html">LICENSE</a>
  </li>
  
</ul>

<h2 id="classes">Classes/Modules</h2>
<ul>
  <li class="module">
    <a href="Bio.html">Bio</a>
  </li>
    <li class="module">
    <a href="Bio/DB.html">Bio::DB</a>
  </li>
    <li class="class">
    <a href="Bio/DB/Pileup.html">Bio::DB::Pileup</a>
  </li>
    <li class="class">
    <a href="Bio/DB/Vcf.html">Bio::DB::Vcf</a>
  </li>
    <li class="class">
    <a href="Bio/Util.html">Bio::Util</a>
  </li>
    <li class="class">
    <a href="Bio/Util/Gngm.html">Bio::Util::Gngm</a>

    <img class="toc-toggle" src="images/transparent.png" alt="" title="toggle headings">
    <ul class="initially-hidden">
      <li><a href="Bio/Util/Gngm.html#label-Background">Background</a>
      <li><a href="Bio/Util/Gngm.html#label-Example">Example</a>
      <li><a href="Bio/Util/Gngm.html#label-Polymorphisms+and+statistics">Polymorphisms and statistics</a>
      <li><a href="Bio/Util/Gngm.html#label-SNPs">SNPs</a>
      <li><a href="Bio/Util/Gngm.html#label-Short+INDELS">Short INDELS</a>
      <li><a href="Bio/Util/Gngm.html#label-Insertion+Size">Insertion Size</a>
      <li><a href="Bio/Util/Gngm.html#label-Unmapped+Mate+Pairs+%2F+Paired+Ends.">Unmapped Mate Pairs / Paired Ends.</a>
      <li><a href="Bio/Util/Gngm.html#label-Input+types">Input types</a>
      <li><a href="Bio/Util/Gngm.html#label-Workflow">Workflow</a>
      <li><a href="Bio/Util/Gngm.html#label-Prerequisites">Prerequisites</a>
      <li><a href="Bio/Util/Gngm.html#label-Acknowledgements">Acknowledgements</a>
      <li><a href="Bio/Util/Gngm.html#label-Using+bio-gngm">Using bio-gngm</a>
      <li><a href="Bio/Util/Gngm.html#label-API">API</a>
    </ul>
  </li>
  
</ul>

<h2 id="methods">Methods</h2>
<ul>
  
    <li class="method"><a href="Bio/Util/Gngm.html#method-c-new">::new &mdash; Bio::Util::Gngm</a>
  
    <li class="method"><a href="Bio/DB/Vcf.html#method-i-alternatives">#alternatives &mdash; Bio::DB::Vcf</a>
  
    <li class="method"><a href="Bio/Util/Gngm.html#method-i-calculate_clusters">#calculate_clusters &mdash; Bio::Util::Gngm</a>
  
    <li class="method"><a href="Bio/Util/Gngm.html#method-i-calculate_densities">#calculate_densities &mdash; Bio::Util::Gngm</a>
  
    <li class="method"><a href="Bio/Util/Gngm.html#method-i-calculate_signal">#calculate_signal &mdash; Bio::Util::Gngm</a>
  
    <li class="method"><a href="Bio/Util/Gngm.html#method-i-close">#close &mdash; Bio::Util::Gngm</a>
  
    <li class="method"><a href="Bio/Util/Gngm.html#method-i-clusters">#clusters &mdash; Bio::Util::Gngm</a>
  
    <li class="method"><a href="Bio/Util/Gngm.html#method-i-collect_threads">#collect_threads &mdash; Bio::Util::Gngm</a>
  
    <li class="method"><a href="Bio/Util/Gngm.html#method-i-densities">#densities &mdash; Bio::Util::Gngm</a>
  
    <li class="method"><a href="Bio/DB/Pileup.html#method-i-discordant_chastity">#discordant_chastity &mdash; Bio::DB::Pileup</a>
  
    <li class="method"><a href="Bio/Util/Gngm.html#method-i-draw_bands">#draw_bands &mdash; Bio::Util::Gngm</a>
  
    <li class="method"><a href="Bio/Util/Gngm.html#method-i-draw_hit_count">#draw_hit_count &mdash; Bio::Util::Gngm</a>
  
    <li class="method"><a href="Bio/Util/Gngm.html#method-i-draw_peaks">#draw_peaks &mdash; Bio::Util::Gngm</a>
  
    <li class="method"><a href="Bio/Util/Gngm.html#method-i-draw_signal">#draw_signal &mdash; Bio::Util::Gngm</a>
  
    <li class="method"><a href="Bio/Util/Gngm.html#method-i-draw_threads">#draw_threads &mdash; Bio::Util::Gngm</a>
  
    <li class="method"><a href="Bio/Util/Gngm.html#method-i-frequency_histogram">#frequency_histogram &mdash; Bio::Util::Gngm</a>
  
    <li class="method"><a href="Bio/Util/Gngm.html#method-i-get_band">#get_band &mdash; Bio::Util::Gngm</a>
  
    <li class="method"><a href="Bio/Util/Gngm.html#method-i-get_insert_size_frequency">#get_insert_size_frequency &mdash; Bio::Util::Gngm</a>
  
    <li class="method"><a href="Bio/Util/Gngm.html#method-i-get_unmapped_mate_frequency">#get_unmapped_mate_frequency &mdash; Bio::Util::Gngm</a>
  
    <li class="method"><a href="Bio/DB/Vcf.html#method-i-gq">#gq &mdash; Bio::DB::Vcf</a>
  
    <li class="method"><a href="Bio/DB/Vcf.html#method-i-has_just_one_variant-3F">#has_just_one_variant? &mdash; Bio::DB::Vcf</a>
  
    <li class="method"><a href="Bio/Util/Gngm.html#method-i-hit_count">#hit_count &mdash; Bio::Util::Gngm</a>
  
    <li class="method"><a href="Bio/DB/Vcf.html#method-i-is_deletion-3F">#is_deletion? &mdash; Bio::DB::Vcf</a>
  
    <li class="method"><a href="Bio/DB/Vcf.html#method-i-is_indel-3F">#is_indel? &mdash; Bio::DB::Vcf</a>
  
    <li class="method"><a href="Bio/DB/Vcf.html#method-i-is_insertion-3F">#is_insertion? &mdash; Bio::DB::Vcf</a>
  
    <li class="method"><a href="Bio/DB/Pileup.html#method-i-is_snp-3F">#is_snp? &mdash; Bio::DB::Pileup</a>
  
    <li class="method"><a href="Bio/DB/Vcf.html#method-i-mq">#mq &mdash; Bio::DB::Vcf</a>
  
    <li class="method"><a href="Bio/DB/Vcf.html#method-i-non_ref_allele_count">#non_ref_allele_count &mdash; Bio::DB::Vcf</a>
  
    <li class="method"><a href="Bio/DB/Vcf.html#method-i-non_ref_allele_freq">#non_ref_allele_freq &mdash; Bio::DB::Vcf</a>
  
    <li class="method"><a href="Bio/DB/Vcf.html#method-i-pass_quality-3F">#pass_quality? &mdash; Bio::DB::Vcf</a>
  
    <li class="method"><a href="Bio/Util/Gngm.html#method-i-peaks">#peaks &mdash; Bio::Util::Gngm</a>
  
    <li class="method"><a href="Bio/DB/Vcf.html#method-i-pl">#pl &mdash; Bio::DB::Vcf</a>
  
    <li class="method"><a href="Bio/Util/Gngm.html#method-i-signal">#signal &mdash; Bio::Util::Gngm</a>
  
    <li class="method"><a href="Bio/Util/Gngm.html#method-i-snp_positions">#snp_positions &mdash; Bio::Util::Gngm</a>
  
    <li class="method"><a href="Bio/Util/Gngm.html#method-i-threads">#threads &mdash; Bio::Util::Gngm</a>
  
    <li class="method"><a href="Bio/DB/Vcf.html#method-i-to_s">#to_s &mdash; Bio::DB::Vcf</a>
  
    <li class="method"><a href="Bio/DB/Vcf.html#method-i-used_depth">#used_depth &mdash; Bio::DB::Vcf</a>
  
    <li class="method"><a href="Bio/DB/Vcf.html#method-i-variant-3F">#variant? &mdash; Bio::DB::Vcf</a>
  
</ul>


<footer id="validator-badges">
  <p><a href="http://validator.w3.org/check/referer">[Validate]</a>
  <p>Generated by <a href="https://github.com/rdoc/rdoc">RDoc</a> 3.11.
  <p>Generated with the <a href="http://deveiate.org/projects/Darkfish-Rdoc/">Darkfish Rdoc Generator</a> 3.
</footer>

