<!DOCTYPE html>

<html>
<head>
<meta content="text/html; charset=UTF-8" http-equiv="Content-Type">

<title>class Bio::DB::Vcf - RDoc Documentation</title>

<link type="text/css" media="screen" href="../../rdoc.css" rel="stylesheet">

<script type="text/javascript">
  var rdoc_rel_prefix = "../../";
</script>

<script type="text/javascript" charset="utf-8" src="../../js/jquery.js"></script>
<script type="text/javascript" charset="utf-8" src="../../js/navigation.js"></script>
<script type="text/javascript" charset="utf-8" src="../../js/search_index.js"></script>
<script type="text/javascript" charset="utf-8" src="../../js/search.js"></script>
<script type="text/javascript" charset="utf-8" src="../../js/searcher.js"></script>
<script type="text/javascript" charset="utf-8" src="../../js/darkfish.js"></script>


<body id="top" class="class">
<nav id="metadata">
  <nav id="home-section" class="section">
  <h3 class="section-header">
    <a href="../../index.html">Home</a>
    <a href="../../table_of_contents.html#classes">Classes</a>
    <a href="../../table_of_contents.html#methods">Methods</a>
  </h3>
</nav>


  <nav id="search-section" class="section project-section" class="initially-hidden">
  <form action="#" method="get" accept-charset="utf-8">
    <h3 class="section-header">
      <input type="text" name="search" placeholder="Search" id="search-field"
             title="Type to search, Up and Down to navigate, Enter to load">
    </h3>
  </form>

  <ul id="search-results" class="initially-hidden"></ul>
</nav>


  <div id="file-metadata">
    <nav id="file-list-section" class="section">
  <h3 class="section-header">Defined In</h3>
  <ul>
    <li>lib/bio/util/bio-gngm.rb
  </ul>
</nav>

    
  </div>

  <div id="class-metadata">
    
    <nav id="parent-class-section" class="section">
  <h3 class="section-header">Parent</h3>
  
  <p class="link">Object
  
</nav>

    
    <!-- Method Quickref -->
<nav id="method-list-section" class="section">
  <h3 class="section-header">Methods</h3>

  <ul class="link-list">
    
    <li><a href="#method-i-alternatives">#alternatives</a>
    
    <li><a href="#method-i-gq">#gq</a>
    
    <li><a href="#method-i-is_indel-3F">#is_indel?</a>
    
    <li><a href="#method-i-is_mnp-3F">#is_mnp?</a>
    
    <li><a href="#method-i-is_snp-3F">#is_snp?</a>
    
    <li><a href="#method-i-mq">#mq</a>
    
    <li><a href="#method-i-non_ref_allele_count">#non_ref_allele_count</a>
    
    <li><a href="#method-i-non_ref_allele_freq">#non_ref_allele_freq</a>
    
    <li><a href="#method-i-pass_quality-3F">#pass_quality?</a>
    
    <li><a href="#method-i-pl">#pl</a>
    
    <li><a href="#method-i-to_s">#to_s</a>
    
    <li><a href="#method-i-used_depth">#used_depth</a>
    
    <li><a href="#method-i-variant-3F">#variant?</a>
    
  </ul>
</nav>

  </div>

  <div id="project-metadata">
    <nav id="fileindex-section" class="section project-section">
  <h3 class="section-header">Pages</h3>

  <ul>
  
    <li class="file"><a href="../../LICENSE_txt.html">LICENSE</a>
  
  </ul>
</nav>

    <nav id="classindex-section" class="section project-section">
  <h3 class="section-header">Class and Module Index</h3>

  <ul class="link-list">
  
    <li><a href="../../Bio.html">Bio</a>
  
    <li><a href="../../Bio/DB.html">Bio::DB</a>
  
    <li><a href="../../Bio/DB/FastaLengthDB.html">Bio::DB::FastaLengthDB</a>
  
    <li><a href="../../Bio/DB/Pileup.html">Bio::DB::Pileup</a>
  
    <li><a href="../../Bio/DB/Vcf.html">Bio::DB::Vcf</a>
  
    <li><a href="../../Bio/Util.html">Bio::Util</a>
  
    <li><a href="../../Bio/Util/Gngm.html">Bio::Util::Gngm</a>
  
  </ul>
</nav>

  </div>
</nav>

<div id="documentation">
  <h1 class="class">class Bio::DB::Vcf</h1>

  <div id="description" class="description">
    
<p>Extends the methods of the <a href="Vcf.html">Bio::DB::Vcf</a> class in
bio-samtools. A <a href="Vcf.html">Vcf</a> object represents the VCF format
described at  <a
href="http://www.1000genomes.org/node/101">www.1000genomes.org/node/101</a>
. The <a href="Vcf.html">Bio::DB::Vcf</a> object returns all information in
the VCF line, but the implementation here acts as if there is only possibly
one variant at each position and ignores positions at which there may be
multiple variants. <a href="Vcf.html">Vcf</a> format is only used when the
<a href="../Util/Gngm.html">Bio::Util::Gngm</a> object requests information
about indels using SAMtools mpileup method.</p>

  </div><!-- description -->

  
  
  
  <section id="5Buntitled-5D" class="documentation-section">
    

    

    

    

    <!-- Methods -->
    
     <section id="public-instance-5Buntitled-5D-method-details" class="method-section section">
      <h3 class="section-header">Public Instance Methods</h3>

    
      <div id="method-i-alternatives" class="method-detail ">
        
        <div class="method-heading">
          <span class="method-name">alternatives</span><span
            class="method-args">()</span>
          <span class="method-click-advice">click to toggle source</span>
        </div>
        

        <div class="method-description">
          
          <p>List of alternate alleles at this locus, obtained by splitting the vcf.alt
attribute string on commas</p>

<p>Example vcf.alt = “ACT,TCA”</p>

<pre>vcf.alternatives = [&quot;ACT&quot;, &quot;TCA&quot;]</pre>

<p>vcf.alt = “T”</p>

<pre>vcf.alternatives = [&quot;T&quot;]</pre>
          

          
          <div class="method-source-code" id="alternatives-source">
            <pre><span class="ruby-comment"># File lib/bio/util/bio-gngm.rb, line 123</span>
<span class="ruby-keyword">def</span> <span class="ruby-identifier">alternatives</span>
  <span class="ruby-keyword">self</span>.<span class="ruby-identifier">alt</span>.<span class="ruby-identifier">split</span>(<span class="ruby-string">&quot;,&quot;</span>) <span class="ruby-keyword">rescue</span> []
<span class="ruby-keyword">end</span></pre>
          </div><!-- alternatives-source -->
          
        </div>

        

        
      </div><!-- alternatives-method -->

    
      <div id="method-i-gq" class="method-detail ">
        
        <div class="method-heading">
          <span class="method-name">gq</span><span
            class="method-args">()</span>
          <span class="method-click-advice">click to toggle source</span>
        </div>
        

        <div class="method-description">
          
          <p>Returns the genotype quality score from the sample data (as defined by the
<a href="Vcf.html">Vcf</a> GQ attribute) for the first sample in the <a
href="Vcf.html">Vcf</a> only.</p>
          

          
          <div class="method-source-code" id="gq-source">
            <pre><span class="ruby-comment"># File lib/bio/util/bio-gngm.rb, line 146</span>
<span class="ruby-keyword">def</span> <span class="ruby-identifier">gq</span> 
  <span class="ruby-keyword">self</span>.<span class="ruby-identifier">samples</span>[<span class="ruby-string">&quot;1&quot;</span>][<span class="ruby-string">&quot;GQ&quot;</span>].<span class="ruby-identifier">to_f</span> <span class="ruby-keyword">rescue</span> <span class="ruby-value">0.0</span>
<span class="ruby-keyword">end</span></pre>
          </div><!-- gq-source -->
          
        </div>

        

        
      </div><!-- gq-method -->

    
      <div id="method-i-is_indel-3F" class="method-detail ">
        
        <div class="method-heading">
          <span class="method-name">is_indel?</span><span
            class="method-args">(options)</span>
          <span class="method-click-advice">click to toggle source</span>
        </div>
        

        <div class="method-description">
          
          <p>Returns true if ref col is different in length from any of the entries in
alt column</p>
          

          
          <div class="method-source-code" id="is_indel-3F-source">
            <pre><span class="ruby-comment"># File lib/bio/util/bio-gngm.rb, line 186</span>
<span class="ruby-keyword">def</span> <span class="ruby-identifier">is_indel?</span>(<span class="ruby-identifier">options</span>)
  <span class="ruby-keyword">return</span> <span class="ruby-keyword">true</span> <span class="ruby-keyword">if</span> <span class="ruby-keyword">self</span>.<span class="ruby-identifier">variant?</span> <span class="ruby-keyword">and</span> <span class="ruby-keyword">self</span>.<span class="ruby-identifier">alternatives</span>.<span class="ruby-identifier">any?</span> {<span class="ruby-operator">|</span><span class="ruby-identifier">x</span><span class="ruby-operator">|</span> <span class="ruby-identifier">x</span>.<span class="ruby-identifier">length</span> <span class="ruby-operator">!=</span> <span class="ruby-keyword">self</span>.<span class="ruby-identifier">ref</span>.<span class="ruby-identifier">length</span>} <span class="ruby-keyword">and</span> <span class="ruby-keyword">self</span>.<span class="ruby-identifier">pass_quality?</span>(<span class="ruby-identifier">options</span>)
  <span class="ruby-keyword">false</span>
<span class="ruby-keyword">end</span></pre>
          </div><!-- is_indel-3F-source -->
          
        </div>

        

        
      </div><!-- is_indel-3F-method -->

    
      <div id="method-i-is_mnp-3F" class="method-detail ">
        
        <div class="method-heading">
          <span class="method-name">is_mnp?</span><span
            class="method-args">(options)</span>
          <span class="method-click-advice">click to toggle source</span>
        </div>
        

        <div class="method-description">
          
          <p>returns true if ref col has same length as all alternatives and position
variant passes quality</p>
          

          
          <div class="method-source-code" id="is_mnp-3F-source">
            <pre><span class="ruby-comment"># File lib/bio/util/bio-gngm.rb, line 174</span>
<span class="ruby-keyword">def</span> <span class="ruby-identifier">is_mnp?</span>(<span class="ruby-identifier">options</span>)
  <span class="ruby-keyword">return</span> <span class="ruby-keyword">true</span> <span class="ruby-keyword">if</span> <span class="ruby-keyword">self</span>.<span class="ruby-identifier">alternatives</span>.<span class="ruby-identifier">all?</span> {<span class="ruby-operator">|</span><span class="ruby-identifier">x</span><span class="ruby-operator">|</span> <span class="ruby-identifier">x</span>.<span class="ruby-identifier">length</span> <span class="ruby-operator">==</span> <span class="ruby-keyword">self</span>.<span class="ruby-identifier">ref</span>.<span class="ruby-identifier">length</span>} <span class="ruby-keyword">and</span> <span class="ruby-keyword">self</span>.<span class="ruby-identifier">pass_quality?</span>(<span class="ruby-identifier">options</span>)
  <span class="ruby-keyword">false</span>
<span class="ruby-keyword">end</span></pre>
          </div><!-- is_mnp-3F-source -->
          
        </div>

        

        
      </div><!-- is_mnp-3F-method -->

    
      <div id="method-i-is_snp-3F" class="method-detail ">
        
        <div class="method-heading">
          <span class="method-name">is_snp?</span><span
            class="method-args">(options)</span>
          <span class="method-click-advice">click to toggle source</span>
        </div>
        

        <div class="method-description">
          
          <p>returns true if ref col has length of 1 and is_mnp?</p>
          

          
          <div class="method-source-code" id="is_snp-3F-source">
            <pre><span class="ruby-comment"># File lib/bio/util/bio-gngm.rb, line 180</span>
<span class="ruby-keyword">def</span> <span class="ruby-identifier">is_snp?</span>(<span class="ruby-identifier">options</span>)
  <span class="ruby-keyword">return</span> <span class="ruby-keyword">true</span> <span class="ruby-keyword">if</span> <span class="ruby-keyword">self</span>.<span class="ruby-identifier">is_mnp?</span>(<span class="ruby-identifier">options</span>) <span class="ruby-keyword">and</span> <span class="ruby-keyword">self</span>.<span class="ruby-identifier">ref</span>.<span class="ruby-identifier">length</span> <span class="ruby-operator">==</span> <span class="ruby-value">1</span>
  <span class="ruby-keyword">false</span>
<span class="ruby-keyword">end</span></pre>
          </div><!-- is_snp-3F-source -->
          
        </div>

        

        
      </div><!-- is_snp-3F-method -->

    
      <div id="method-i-mq" class="method-detail ">
        
        <div class="method-heading">
          <span class="method-name">mq</span><span
            class="method-args">()</span>
          <span class="method-click-advice">click to toggle source</span>
        </div>
        

        <div class="method-description">
          
          <p>Returns the mean Mapping Quality from the reads over this position as
defined by the <a href="Vcf.html">Vcf</a> MQ attribute.</p>
          

          
          <div class="method-source-code" id="mq-source">
            <pre><span class="ruby-comment"># File lib/bio/util/bio-gngm.rb, line 141</span>
<span class="ruby-keyword">def</span> <span class="ruby-identifier">mq</span>
  <span class="ruby-keyword">self</span>.<span class="ruby-identifier">info</span>[<span class="ruby-string">&quot;MQ&quot;</span>].<span class="ruby-identifier">to_f</span> <span class="ruby-keyword">rescue</span> <span class="ruby-value">0.0</span>
<span class="ruby-keyword">end</span></pre>
          </div><!-- mq-source -->
          
        </div>

        

        
      </div><!-- mq-method -->

    
      <div id="method-i-non_ref_allele_count" class="method-detail ">
        
        <div class="method-heading">
          <span class="method-name">non_ref_allele_count</span><span
            class="method-args">()</span>
          <span class="method-click-advice">click to toggle source</span>
        </div>
        

        <div class="method-description">
          
          <p>Returns the depth of reads containing the non reference allele. IE the sum
of the last two figures in the DP4 attribute.</p>
          

          
          <div class="method-source-code" id="non_ref_allele_count-source">
            <pre><span class="ruby-comment"># File lib/bio/util/bio-gngm.rb, line 128</span>
<span class="ruby-keyword">def</span> <span class="ruby-identifier">non_ref_allele_count</span>
  <span class="ruby-keyword">self</span>.<span class="ruby-identifier">info</span>[<span class="ruby-string">&quot;DP4&quot;</span>].<span class="ruby-identifier">split</span>(<span class="ruby-string">&quot;,&quot;</span>)[<span class="ruby-value">2</span><span class="ruby-operator">..</span><span class="ruby-value">3</span>].<span class="ruby-identifier">inject</span> {<span class="ruby-operator">|</span><span class="ruby-identifier">sum</span>,<span class="ruby-identifier">n</span><span class="ruby-operator">|</span> <span class="ruby-identifier">sum</span>.<span class="ruby-identifier">to_f</span> <span class="ruby-operator">+</span> <span class="ruby-identifier">n</span>.<span class="ruby-identifier">to_f</span> } <span class="ruby-keyword">rescue</span> <span class="ruby-value">0.0</span>
<span class="ruby-keyword">end</span></pre>
          </div><!-- non_ref_allele_count-source -->
          
        </div>

        

        
      </div><!-- non_ref_allele_count-method -->

    
      <div id="method-i-non_ref_allele_freq" class="method-detail ">
        
        <div class="method-heading">
          <span class="method-name">non_ref_allele_freq</span><span
            class="method-args">()</span>
          <span class="method-click-advice">click to toggle source</span>
        </div>
        

        <div class="method-description">
          
          <p>Returns the non-reference allele frequency based on depth of reads used for
the genotype call,</p>

<p>IE</p>

<pre>vcf.non_ref_allele_count / vcf.used_depth</pre>
          

          
          <div class="method-source-code" id="non_ref_allele_freq-source">
            <pre><span class="ruby-comment"># File lib/bio/util/bio-gngm.rb, line 136</span>
<span class="ruby-keyword">def</span> <span class="ruby-identifier">non_ref_allele_freq</span>
  <span class="ruby-keyword">self</span>.<span class="ruby-identifier">non_ref_allele_count</span> <span class="ruby-operator">/</span> <span class="ruby-keyword">self</span>.<span class="ruby-identifier">used_depth</span>
<span class="ruby-keyword">end</span></pre>
          </div><!-- non_ref_allele_freq-source -->
          
        </div>

        

        
      </div><!-- non_ref_allele_freq-method -->

    
      <div id="method-i-pass_quality-3F" class="method-detail ">
        
        <div class="method-heading">
          <span class="method-name">pass_quality?</span><span
            class="method-args">(options)</span>
          <span class="method-click-advice">click to toggle source</span>
        </div>
        

        <div class="method-description">
          
          <p>Returns true if the position passes criteria</p>

<p>Options and Defaults:</p>
<ul><li>
<p>:min_depth =&gt; 2</p>
</li><li>
<p>:min_non_ref_count =&gt; 2</p>
</li><li>
<p>:mapping_quality =&gt; 10</p>
</li></ul>

<p>Example</p>

<pre class="ruby"><span class="ruby-identifier">vcf</span>.<span class="ruby-identifier">pass_quality?</span>(:<span class="ruby-identifier">min_depth</span> =<span class="ruby-operator">&gt;</span> <span class="ruby-value">5</span>, :<span class="ruby-identifier">min_non_ref_count</span> =<span class="ruby-operator">&gt;</span> <span class="ruby-value">2</span>, :<span class="ruby-identifier">mapping_quality</span> =<span class="ruby-operator">&gt;</span> <span class="ruby-value">25</span>, :<span class="ruby-identifier">min_snp_quality</span> =<span class="ruby-operator">&gt;</span> <span class="ruby-value">20</span>)
</pre>
          

          
          <div class="method-source-code" id="pass_quality-3F-source">
            <pre><span class="ruby-comment"># File lib/bio/util/bio-gngm.rb, line 169</span>
<span class="ruby-keyword">def</span> <span class="ruby-identifier">pass_quality?</span>(<span class="ruby-identifier">options</span>)
  (<span class="ruby-keyword">self</span>.<span class="ruby-identifier">used_depth</span> <span class="ruby-operator">&gt;=</span> <span class="ruby-identifier">options</span>[<span class="ruby-value">:min_depth</span>] <span class="ruby-keyword">and</span> <span class="ruby-keyword">self</span>.<span class="ruby-identifier">mq</span> <span class="ruby-operator">&gt;=</span> <span class="ruby-identifier">options</span>[<span class="ruby-value">:mapping_quality</span>] <span class="ruby-keyword">and</span> <span class="ruby-keyword">self</span>.<span class="ruby-identifier">non_ref_allele_count</span> <span class="ruby-operator">&gt;=</span> <span class="ruby-identifier">options</span>[<span class="ruby-value">:min_non_ref_count</span>] <span class="ruby-keyword">and</span> <span class="ruby-keyword">self</span>.<span class="ruby-identifier">qual</span> <span class="ruby-operator">&gt;=</span> <span class="ruby-identifier">options</span>[<span class="ruby-value">:min_snp_quality</span>])
<span class="ruby-keyword">end</span></pre>
          </div><!-- pass_quality-3F-source -->
          
        </div>

        

        
      </div><!-- pass_quality-3F-method -->

    
      <div id="method-i-pl" class="method-detail ">
        
        <div class="method-heading">
          <span class="method-name">pl</span><span
            class="method-args">()</span>
          <span class="method-click-advice">click to toggle source</span>
        </div>
        

        <div class="method-description">
          
          <p>Returns the phred scaled likelihood of the first non-reference allele (as
defined by the <a href="Vcf.html">Vcf</a> PL attribute) for the first
sample in the <a href="Vcf.html">Vcf</a> only.</p>
          

          
          <div class="method-source-code" id="pl-source">
            <pre><span class="ruby-comment"># File lib/bio/util/bio-gngm.rb, line 151</span>
<span class="ruby-keyword">def</span> <span class="ruby-identifier">pl</span>
  <span class="ruby-keyword">self</span>.<span class="ruby-identifier">samples</span>[<span class="ruby-string">&quot;1&quot;</span>][<span class="ruby-string">&quot;PL&quot;</span>].<span class="ruby-identifier">split</span>(<span class="ruby-string">&quot;,&quot;</span>)[<span class="ruby-value">1</span>].<span class="ruby-identifier">to_f</span> <span class="ruby-keyword">rescue</span> <span class="ruby-value">0.0</span>
<span class="ruby-keyword">end</span></pre>
          </div><!-- pl-source -->
          
        </div>

        

        
      </div><!-- pl-method -->

    
      <div id="method-i-to_s" class="method-detail ">
        
        <div class="method-heading">
          <span class="method-name">to_s</span><span
            class="method-args">()</span>
          <span class="method-click-advice">click to toggle source</span>
        </div>
        

        <div class="method-description">
          
          <p>Return a short string representing chromosome, position, reference
sequence, alt sequence(s) and the info string of the <a
href="Vcf.html">Vcf</a> object.</p>
          

          
          <div class="method-source-code" id="to_s-source">
            <pre><span class="ruby-comment"># File lib/bio/util/bio-gngm.rb, line 107</span>
<span class="ruby-keyword">def</span> <span class="ruby-identifier">to_s</span>
  <span class="ruby-node">&quot;#{self.chrom} #{self.pos} #{self.ref} #{self.alt} #{self.info}&quot;</span>
<span class="ruby-keyword">end</span></pre>
          </div><!-- to_s-source -->
          
        </div>

        

        
      </div><!-- to_s-method -->

    
      <div id="method-i-used_depth" class="method-detail ">
        
        <div class="method-heading">
          <span class="method-name">used_depth</span><span
            class="method-args">()</span>
          <span class="method-click-advice">click to toggle source</span>
        </div>
        

        <div class="method-description">
          
          <p>The depth of reads actually used in the genotype call by Vcftools. The sum
of the DP4 attribute. Returns 0.0 if no value is calculated.</p>
          

          
          <div class="method-source-code" id="used_depth-source">
            <pre><span class="ruby-comment"># File lib/bio/util/bio-gngm.rb, line 112</span>
<span class="ruby-keyword">def</span> <span class="ruby-identifier">used_depth</span>
  <span class="ruby-keyword">self</span>.<span class="ruby-identifier">info</span>[<span class="ruby-string">&quot;DP4&quot;</span>].<span class="ruby-identifier">split</span>(<span class="ruby-string">&quot;,&quot;</span>).<span class="ruby-identifier">inject</span> {<span class="ruby-operator">|</span><span class="ruby-identifier">sum</span>,<span class="ruby-identifier">n</span><span class="ruby-operator">|</span> <span class="ruby-identifier">sum</span>.<span class="ruby-identifier">to_f</span> <span class="ruby-operator">+</span> <span class="ruby-identifier">n</span>.<span class="ruby-identifier">to_f</span>} <span class="ruby-keyword">rescue</span> <span class="ruby-value">0.0</span>
<span class="ruby-keyword">end</span></pre>
          </div><!-- used_depth-source -->
          
        </div>

        

        
      </div><!-- used_depth-method -->

    
      <div id="method-i-variant-3F" class="method-detail ">
        
        <div class="method-heading">
          <span class="method-name">variant?</span><span
            class="method-args">()</span>
          <span class="method-click-advice">click to toggle source</span>
        </div>
        

        <div class="method-description">
          
          <p>returns true if the <code>alt</code> column of the <a
href="Vcf.html">Vcf</a> is not <strong>.</strong></p>

<p>Examples</p>

<p>vcf record = 20     14370   rs6054257 G      A       29   PASS …</p>

<pre class="ruby"><span class="ruby-identifier">vcf</span>.<span class="ruby-identifier">variant?</span> <span class="ruby-comment">#=&gt; true</span>
</pre>

<p>vcf record = 20     1230237 .         T      .       47   PASS …</p>

<pre class="ruby"><span class="ruby-identifier">vcf</span>.<span class="ruby-identifier">variant?</span> <span class="ruby-comment">#=&gt; false</span>
</pre>
          

          
          <div class="method-source-code" id="variant-3F-source">
            <pre><span class="ruby-comment"># File lib/bio/util/bio-gngm.rb, line 102</span>
<span class="ruby-keyword">def</span> <span class="ruby-identifier">variant?</span>
  <span class="ruby-keyword">not</span> <span class="ruby-keyword">self</span>.<span class="ruby-identifier">alt</span> <span class="ruby-operator">==</span> <span class="ruby-string">&quot;.&quot;</span> <span class="ruby-keyword">rescue</span> <span class="ruby-keyword">false</span>
<span class="ruby-keyword">end</span></pre>
          </div><!-- variant-3F-source -->
          
        </div>

        

        
      </div><!-- variant-3F-method -->

    
    </section><!-- public-instance-method-details -->
  
  </section><!-- 5Buntitled-5D -->

</div><!-- documentation -->


<footer id="validator-badges">
  <p><a href="http://validator.w3.org/check/referer">[Validate]</a>
  <p>Generated by <a href="https://github.com/rdoc/rdoc">RDoc</a> 3.12.
  <p>Generated with the <a href="http://deveiate.org/projects/Darkfish-Rdoc/">Darkfish Rdoc Generator</a> 3.
</footer>

